### ISSUE TITLE
Relative links to `./[2]abc.md` crash the app

### ISSUE THREAD

**Issue by mikenikles** [vitebook/vitebook#45]
Hey :wave:,

I have the following directories & files:

![image](https://user-images.githubusercontent.com/788827/155911593-08a831e0-f03a-4902-a8d5-92d3937490c8.png)

In `services/docs/src/[1]introduction/[1]what-is-webstone-education.md`, I have a relative link like so:

```markdown
The best part? You get to [choose your own technology stack](./[3]build-your-own-curriculum.md) :tada:!
```

When I load Vitebook, it crashes with the following error:

```
services/docs dev: node:fs:585
services/docs dev:   handleErrorFromBinding(ctx);
services/docs dev:   ^
services/docs dev: Error: ENOENT: no such file or directory, open '/workspace/platform/services/docs/src/[1]introduction/%5B3%5Dbuild-your-own-curriculum.md'
services/docs dev:     at Object.openSync (node:fs:585:3)
services/docs dev:     at Object.readFileSync (node:fs:453:35)
services/docs dev:     at readRecentlyChangedFile (file:///workspace/platform/node_modules/.pnpm/@vitebook+core@0.23.2/node_modules/@vitebook/core/dist/node/index.js:8294:22)
services/docs dev:     at filePathToRoute (file:///workspace/platform/node_modules/.pnpm/@vitebook+core@0.23.2/node_modules/@vitebook/core/dist/node/index.js:8331:30)
services/docs dev:     at Object.parser.renderer.rules.link_open (file:///workspace/platform/node_modules/.pnpm/@vitebook+markdown@0.23.2_fc6df10e6b401eecad6179f4f69ddf69/node_modules/@vitebook/markdown/dist/node/index.js:17063:23)
services/docs dev:     at Renderer.renderInline (file:///workspace/platform/node_modules/.pnpm/@vitebook+markdown@0.23.2_fc6df10e6b401eecad6179f4f69ddf69/node_modules/@vitebook/markdown/dist/node/index.js:1000:32)
services/docs dev:     at Renderer.render (file:///workspace/platform/node_modules/.pnpm/@vitebook+markdown@0.23.2_fc6df10e6b401eecad6179f4f69ddf69/node_modules/@vitebook/markdown/dist/node/index.js:1025:26)
services/docs dev:     at Renderer.parser.renderer.render (file:///workspace/platform/node_modules/.pnpm/@vitebook+markdown@0.23.2_fc6df10e6b401eecad6179f4f69ddf69/node_modules/@vitebook/markdown/dist/node/index.js:17277:25)
services/docs dev:     at MarkdownIt2.render (file:///workspace/platform/node_modules/.pnpm/@vitebook+markdown@0.23.2_fc6df10e6b401eecad6179f4f69ddf69/node_modules/@vitebook/markdown/dist/node/index.js:4746:28)
services/docs dev:     at MarkdownIt2.parser.render (file:///workspace/platform/node_modules/.pnpm/@vitebook+markdown@0.23.2_fc6df10e6b401eecad6179f4f69ddf69/node_modules/@vitebook/markdown/dist/node/index.js:16914:20) {
services/docs dev:   errno: -2,
services/docs dev:   syscall: 'open',
services/docs dev:   code: 'ENOENT',
services/docs dev:   path: '/workspace/platform/services/docs/src/[1]introduction/%5B2%5Dbuild-your-own-curriculum.md'
services/docs dev: }
services/docs dev: Failed
/workspace/platform/services/docs:
 ERR_PNPM_RECURSIVE_RUN_FIRST_FAIL  @webstone/docs@0.0.0 dev: `vitebook dev -p 8080`
Exit status 1
 ELIFECYCLE  Command failed with exit code 1.
gitpod /workspace/platform $ 
```

If I remove the `.md` ending, the page loads but the link is still broken. I'd like to keep the `.md` extension so that links also work when someone views the file in GitHub.

I thought I'd report this here before I start to investigate and will follow up with any findings I may have.

**Comment by mikenikles** [comment]
I only played with that in my `node_modules` code, but here's what I think we need to fix.

## Use `decodeURIComponent` to get the correct internal link
Basically, change [this code](https://github.com/vitebook/vitebook/blob/d5012a2e2c9abf5636f7a0b9eefc44c3dc8d3ae5/packages/markdown/src/node/parser/plugins/linksPlugin.ts#L66) to the following:

```diff
-const internalLinkMatch = hrefLink.match(
+const internalLinkMatch = decodeURIComponent(hrefLink).match(
  /^((?:.*)(?:\/|\.md|\.html))(#.*)?$/,
);
```

## `filePathToRoute` returns a Promise

The `filePathToRoute` function used to by synchronouse, until custom routes were introduced [here](https://github.com/vitebook/vitebook/commit/ffe7b54e0bb7f7b4c179f31c33e9e70453832912#diff-ad9a9cf0b778e5734e498f7ddb5fe715bbd14b9e7b01a944a1c046ce240d7096R140).

Because it now returns a promise, we should `await` it [here](https://github.com/vitebook/vitebook/blob/main/packages/markdown/src/node/parser/plugins/linksPlugin.ts#L100).

However, that would make the `parser.renderer.rules.link_open` function async as well, which I don't think is a valid case.

Curious to see if you have any suggestions on how to tackle the Promise issue.

**Comment by mihar-22** [comment]
That sounds about right to me. It seems more likely the main issue is awaiting the return of `filePathToRoute`. It seems a little odd we have to decode the URI in the `href` attribute since that shouldn't be encoded? Did that resolve something in your findings?

We can make the `filePathToRoute` function synchronous by using `readFileSync` instead of `readFile` from the Node `fs` lib.

**Comment by mikenikles** [comment]
I suspect on of the `markdown-it-*` dependencies encodes the href attribute, although I couldn't find the code when I checked their source. I also don't see any recent dependency upgrades, so I'm a bit at a loss as to how this has worked so far. Keep in mind though, I may very well miss something here :).

The `parser.renderer.rules.link_open` function receives the `link_open` token with the attribute already encoded. Without the explicit `decodeURIComponent(hrefLink)`, we get a `no such file or directory` error trying to open `services/docs/src/[1]introduction/%5B3%5Dbuild-your-own-curriculum.md`

Your suggested changes (plus the `decodeURIComponent` call) do fix the issue I reported here :tada:. Let me fire up a PR and I'll let you review it.
